### ISSUE TITLE
Adjusting constructed molecule torsions to match building blocks

### ISSUE THREAD

**Issue by joshkamm** [JelfsMaterialsGroup/stko#80]
Follow up to #63 and #77. Here's a concrete example of how rdkit's torsions interact with molecule construction using the new torsion result interface.

If I replace the following lines 59-60 in torsion\_example.py

```python
    bb1 = stk.BuildingBlock('NCCNCCN', [stk.PrimaryAminoFactory()])
    bb2 = stk.BuildingBlock('O=CCCC=O', [stk.AldehydeFactory()])
```

with

```python
    bb1 = stk.BuildingBlock('CCCC', [stk.SingleAtom(stk.C(1))])
    bb2 = stk.BuildingBlock('C', [stk.SingleAtom(stk.C(0))])
```

I get the following output (I hope this is consistent and reproducible).

```
BuildingBlock('CCCC', (SingleAtom(C(1)),))
a Torsion(C(0), C(1), C(2), C(3))
Torsion(C(0), C(1), C(2), C(3)) -179.9975585000165 (0, 1, 2, 3)
<ConstructedMolecule at 139658949541840>
a Torsion(C(14), C(1), C(2), C(3))
Torsion(C(14), C(1), C(2), C(3)) 2.166041425964015 (14, 1, 2, 3)
c Torsion(C(14), C(1), C(2), C(3)) None None None
defaultdict(<class 'list'>, {})
(2.166041425964015, None)
```

Note that rdkit arbitrarily chose to include C(14) from the methane as an atom in the only torsion of the constructed molecule and consequently, the torsion info's building\_block and hence building\_block\_torsion are "None" so the mapping between building block torsion and constructed molecule torsion are effectively lost. I think this would be common in larger molecules.

I'm considering whether it would be helpful to add an option to attempt to adjust constructed molecule torsions to try to make more of them map onto building block torsions. This could look like iterating through rdkit's torsion of the constructed molecule, and checking whether the two central atoms correspond to the two central atoms of a building block torsion. If they do, the torsion of the constructed molecule would be adjusted to match so that building\_block\_torsion returns an entry in the building block's torsion results.

I hope I'm thinking about torsions correctly and not missing something. Does this make sense, and would it be helpful for stko?

Todo:

*   [x] Set up a test based on the example above
*   [x] Try an implementation based on the discussion below
*   [x] Write a pull request and move the rest of this todo list there:

*   Should I move get\_torsion\_info\_angles to utilities since if it's helpful in the test?
*   Add flake8 to vscode to help with project consistency
*   Manual check for code and documentation consistency

**Comment by andrewtarzia** [comment]
Hey @joshkamm I see what you mean now. This example is very clear!

Maybe we should define torsions in a different way through rdkit? Does the same issue occur if you search for torsions with `FindAllPathsOfLengthN` instead (https://www.rdkit.org/docs/source/rdkit.Chem.rdmolops.html)?

**Comment by joshkamm** [comment]
Oh good. In my mind with `FindAllPathsOfLengthN` we would get at least one path for each rotatable bond, but we would also get multiple paths for a single rotatable bond as well as paths corresponding to ring torsions. Any paths that map onto building block torsions could be included but I think identifying which paths correspond to new rotatable bonds generated in the construction process could be more challenging.

The advantages of the CalculateTorsionLists you're currently using is that it sorts out which paths are actually valid torsions, and it separates ring torsions which I find convenient. So if I start with those I think I can treat each rotatable bond independently and output one torsion for each one provided by rdkit.

Can you think of anything I would be missing by assuming that the two central atoms of each CalculateTorsionLists torsion (as an order invariant set) uniquely map one-to-one with each rotatable bond? If that's a valid assumption it seems like postprocessing wouldn't get too involved.

**Comment by andrewtarzia** [comment]
Yeah, my assumption is that FindAllPaths would require some post processing to rule out those you are not interested in. Perhaps using both methods (one to find all torsions, the other to find all rotatable bonds (as you suggested)) and then filtering the torsions by their overlap with a rotatable bond?

I do not know that particular code well enough to comment on the assumption you are making - you can always have a flag of num_rotatable_bonds == num_torsions, where num_rotatable_bonds is also calculated by RDKit? Assuming they behave the same, this would flag up edge cases?

**Comment by joshkamm** [comment]
Hmm this is making me realize I may need to think a bit more deeply about what I mean by rotatable bonds and torsions. Maybe if I try to start implementing it'll be more clear what the challenges are and provide something concrete to work with.

**Comment by andrewtarzia** [comment]
Hey @joshkamm - just curious to see how you are going with this?

**Comment by joshkamm** [comment]
I wrote a pytest based on the example above and have been experimenting on an implementation. I just finished debugging a prototype that now passes the test which was fun but I don't have confidence in its robustness yet. When I have the mental capacity I can look back through, and at the very least take care of a few blatant inefficiencies and clean it up.

I'm working in https://github.com/joshkamm/stko/tree/joshkamm/issue80
